website/statistics.py

```diff
@@ -362,3 +498,133 @@ def get_general_class_stats(students):
         "week": {"runs": weekly_successes + weekly_errors, "fails": weekly_errors},
         "total": {"runs": successes + errors, "fails": errors},
     }
+
+
+def find_program_runs_per_level(program_runs_per_level, success_runs_per_level, error_runs_per_level, programs, level):
+    """
+    Finds program runs per level and appends them to lists.
+
+    :param program_runs_per_level: list of lists containing program run information per level
+    :param success_runs_per_level: list of successful program runs per level
+    :param error_runs_per_level: list of unsuccessful program runs per level
+    :param programs: list of dictionaries containing program information
+    :param level: current level
+    """
+    program_runs_per_level.append([])
+    success_runs_per_level.append([])
+    error_runs_per_level.append([])
+    for program in programs:
+        if program['level'] == level:
+            program_runs_per_level[level - 1].append([program['name'], str(program.get("error"))])
+            if str(program.get("error")) == "True":
+                success_runs_per_level[level - 1].append(program['name'])
+            elif str(program.get("error")) == "False":
+                error_runs_per_level[level - 1].append(program['name'])
+
+
+def find_quizzes_per_level(quizzes_per_level, quizzes, level):
+    """
+    Finds the quizzes per level and appends them to a list.
+
+    :param quizzes_per_level: list of lists containing the number of quizzes per level
+    :param quizzes: list of dictionaries containing quiz information
+    :param level: current level
+    :return: updated list of quizzes per level
+    """
+    quizzes_per_level.append([])
+    for quiz_scores in quizzes:
+        if quiz_scores['level'] == level:
+            # if a quiz level is repeated in a different week, you need to handle it differently.
+            if len(quiz_scores["scores"]) == 1:
+                quizzes_per_level[level - 1].append(quiz_scores["scores"][0])
+            else:
+                quizzes_per_level[level - 1] = quiz_scores["scores"]
+
+
+def calc_avg_quizzes_per_level(avg_quizzes_ran_per_level, quizzes_ran_per_level, level):
+    """
+    Calculates the average number of quizzes ran per level and appends it to a list.
+
+    :param avg_quizzes_ran_per_level: list of average quizzes ran per level
+    :param quizzes_ran_per_level: list of lists containing the number of quizzes ran per level
+    :param level: current level
+    :return: updated list of average quizzes ran per level
+    """
+    if not quizzes_ran_per_level[level - 1]:
+        avg_quizzes_ran_per_level.append(0)
+    else:
+        quizzes_per_level = quizzes_ran_per_level[level - 1]
+        avg_quiz = sum(quizzes_per_level) / len(quizzes_per_level)
```

Comment:
All of the calculations look so much better now!! Thank you for changing them :D